Align HeadLogic: require snapshot not too far in the future

ch1bo opened the pull request:
The first of multiple PRs about aligning the HeadLogic with the spec.

* Changes the off-chain protocol logic to be more strict in allowed snapshot numbers on ReqSn and AckSn as specified. Some of our HeadLogic tests were allowing these "too far in the future" snapshots, but the BehaviorSpec seems unimpacted by this change.

* Introduces `LastSeenSnapshot` to cover for an intermediate state where we have completed a snapshot, but not yet received or sent a new ReqSn.

* Adds an explicit `RequireFailed` logic outcome error to identify the "detect cheating" cases. These are currently just logged like other non progress updates of HeadLogic, but we could handle these by closing the head or so.

* Refactors more of the off-chain protocol into CPS-style to align better with the spec.

---

<!-- Tick off or strike-through / remove if not applicable -->
* ~~[ ] CHANGELOG updated~~
* ~~[ ] Documentation updated~~
* [x] Added and/or updated haddocks
* [x] No new TODOs introduced or explained herafter

ch1bo committed:
Add names from the spec to haddocks

ch1bo committed:
Start using a RequireFailed outcome error

This should indicate which and how a 'require' statement from the
protocol logic spec was invalidated.

ch1bo committed:
Provide explanation for difference in snapshot emission

ch1bo committed:
Reject too new snapshots as defined in the spec

This changes the behavior of the Head protocol to be more restrictive.
It only allows snapshot requests for the next-in-line snapshot.

ch1bo committed:
Remove the redundant 'Event' argument on onOpenNetworkReqSn

ch1bo committed:
Refactor onOpenNEtworkReqSn a tiny bit more

ch1bo committed:
Refactor onOpenNetworkAckSn to checks like in spec

No semantic change expected and tests also indicate this.

The require constraint is not unit tested right now.

ch1bo committed:
Switch to if/then/else instead of case guards

ch1bo committed:
Distinguish between NoSeenSnapshot and LastSeenSnapshot

As this affects how the coordinated head state is serialized, this also
requires to regenerate the golden HeadState.

ffakenz commented on hydra-node/test/Hydra/BehaviorSpec.hs:
```diff
@@ -255,30 +255,29 @@ spec = parallel $ do
 
     it "sending two conflicting transactions should lead one being confirmed and one expired" $
       shouldRunInSim $
-        failAfter 2 $ do
```
why did we remove this?

ch1bo replied:
It was the only failAfter and the other tests would block until `veryLong` happens (in simulated time). I also changed `veryLong` to be `oneMonth` in another PR IIRC.

pgrange reviewed (approved):
I appreciate:
* the link in the comments with the names used in the spec
* meaningful and focused commits which make the PR easier to read

To make it perfect I would:
* add a breaking change entry about the API to the changelog
* revert the [Refactor onOpenNEtworkReqSn](https://github.com/input-output-hk/hydra/pull/733/commits/5d32021004480d23a57d8a02482d3eb04e038d58) as I don't see more meaning expressed after than before
* amend [Refactor onOpenNetworkAckSn to checks like in spec](https://github.com/input-output-hk/hydra/pull/733/commits/805cf179cdbfe12c93d88e08c11989715b20aebb) to remove `requireAckSn` which is introduced but not tested

ch1bo commented:
@pgrange 

> add a breaking change entry about the API to the changelog

I was doubting it's a breaking change and indeed, `SeenSnapshot` is not used in `api.yaml`, but only in the `logs.yaml`. I will move it.

> revert the [Refactor onOpenNEtworkReqSn](https://github.com/input-output-hk/hydra/pull/733/commits/5d32021004480d23a57d8a02482d3eb04e038d58) as I don't see more meaning expressed after than before

I would have preferred a comment in that commit. I disagree, I think it aligns much better with the spec and should make it even clear to non-haskellers that it's a similar control flow.

> amend [Refactor onOpenNetworkAckSn to checks like in spec](https://github.com/input-output-hk/hydra/pull/733/commits/805cf179cdbfe12c93d88e08c11989715b20aebb) to remove requireAckSn which is introduced but not tested

It is tested. Not unit tested, but indirectly via `BehaviorSpec`, where the error outcome does not matter.. as long as the snapshot does not get confirmed. I'll keep it.

ch1bo committed:
Move SeenSnapshot from api.yaml -> logs.yaml

It was not used in the API and hence it can be defined in the logs
schema only to make sure it's not part of the user facing API.

ch1bo committed:
Update haddock descriptions of ReqTx,ReqSn and AckSn handlers

Using language which is understandable in isoluation, but also similar
to the protocol flow from the specification.
